### ISSUE TITLE
Params inferred as unknown when they contain methods with recursive param types

### ISSUE THREAD

**Issue by lynndylanhurley** [microsoft/TypeScript#51578]
# Bug Report

In this example using a recursive generic type, the `b` prop of the second `createFoo` is inferred to be `unknown`. 

```ts
type FooConfig<T> = {
  [K in keyof T]: {
    [P in keyof T[K]]: number | ((f: Foo<T>) => number);
  }
}

type Foo<T> = {
  [K in keyof T]: {
    [P in keyof T[K]]: number;
  }
};

const createFoo = <T, TFoo = Foo<T>>(args: FooConfig<T>): TFoo =>  {
  // code...
}

const foo = createFoo({
  a: {
    i: 1,
  },
  b: {
    ii: () => 1, // <-- no function arg, works as expected
  }
});

// works as intended
console.log('@-->foo', foo.b.ii);

const foo2 = createFoo({
  a: {
    i: 1,
  },
  b: {
    ii: (f) => f.a.i += 1, // <-- only difference is that this method has an argument
    // iii: 3 // <-- uncomment this line and it works
  }
});

// type error when function argument is present in config
console.log('@-->foo2', foo2.b.ii);
```

This example has two instances of the `createFoo` method. The first call is not using a function argument for the `b.ii` param and everything works as intended.

In the 2nd `createFoo` call, the `b.ii` param takes a method argument that is a recursive reference to the `createFoo` method argument itself.

The recursive reference type itself works as expected (i.e. the `f` in `b.ii` is typed correctly)

But in the resulting output, references to `foo2.b.ii` cause a type error.

Strangely, if I add a `b.iii` param as either a `number` or a `() => number` type (i.e. not using the recursive reference), then the `foo2.b.ii` type error is resolved.

### 🔎 Search Terms

* `recursive unknown`

### 🕗 Version & Regression Information

This bug is present in every version of TypeScript available in the TS Playground.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nAwnAdgMwJYHMA8AVAPigF4oBvAWACgooBtAaSgxSgGsIQ40o8BdAFzlqtWnQAKzVhy488jPoKgoArgFsARhABOUAD5QAFIbRD4cfAQCUJIqs06rAbhFQAvtQ9VqoSLASWJMI09Ews7JzcvEqUIWKS4TJR8gyKQvZa2i4hXm7Z1ADGqADOwFAF2hAAhsAQ5kH4ADS89aTmlgSGVdpYxWYIyOjYllZCeK1EwbSVwCrarADyGgBWEAXAAHQQKMDaGBDFXT3FVhuVACYqBRDGahBqcM10bM0AblUANnw2xESxolA7g9nlAqsUIrIWgg+EElqt1lsdnsDoZ3h9QeDzINMLhCM9IjwsagcZZvmcIJdrrd7nAAExPNj0qBo2nfWxTAHMHiGXwQKIskjEUgAcjQKhQ6wwqGFNn+nNoQLpz1pGIhyXM+Mh42hMNILMMiucrgBbigEA+xWgcvlitpytVSTkGsdULgiiCArBrs16uhPqdCH9rsU2XlXnlMzmrFtodEbmaZFNXu1cCDKe++TiUEj80BNNj8fISfBKaNVC8hRKZTQCCCFWqtXMhjlVSE1uYQgAjI1XPHXBo28bmBghIYfkRu1AAPRTqA4AC08+UcCgYolwClrG6WGaAHc4No2OCvRAAB6QdYU3ueMvUGdQfeH4-gli1FDnK9UIooYpwD4QDYPjgLBDGFAABRcCBrOBhWaaCNg0DYMAwW8vyrVcEBVUh6xqOoEGbVxWw5WgRygbtex7EIB2I4dRzQcdVw2KokKgABqUhJ3vBcl1QD4QCgc4MDQNAdG2a5mHBYAAAsaigaSMHBO5pLgc4oBk48tx6dRtmAId72Q0iAGZp1nbioHFIo1DuHY5KkhSoA+FhoCqd9mDKR8j2vctUPvXkzW0bQDwfKTtlXCyN1QUEtOssp7LASpLRs8JvxxSsfz-ACgJA8DIOg2lYIwukEKQlDsiAA)

### 💻 Code

See above

### 🙁 Actual behavior

The `foo2` instance results in the following `.d.ts`:

```ts
declare const foo2: Foo<{
    a: {
        i: unknown;
    };
    b: unknown;
}>;
```

### 🙂 Expected behavior

The `foo2` instance should result in something like this:

```ts
declare const foo2: Foo<{
    a: {
        i: unknown;
    };
    b: {
        ii: unknown;
    };
}>;
```

**Comment by lynndylanhurley** [comment]
I've tried for several days to get help on the typescript discord channel, but I haven't received a single response. So I'm wondering if this is maybe a bug.

Also posted this to [StackOverflow](https://stackoverflow.com/questions/74481388/why-are-nested-properties-of-this-typescript-recursive-function-argument-inferre).

**Comment by RyanCavanaugh** [comment]
See #47599

**Comment by lynndylanhurley** [comment]
Thank you @RyanCavanaugh. That issue describes my situation perfectly.

Am I correct in assuming that this input type should be inferrable?

```ts
const foo2 = createFoo({
  a: {
    i: 1,
  },
  b: {
    ii: (f) => 1 // f is a recursive reference to the base createFoo method param
  }
});
```

Or am I trying to do something that goes against the design goals of TypeScript?

**Comment by RyanCavanaugh** [comment]
I don't see any theoretical blocker to inferring that.

That said, I think *in general* an algorithm that would be capable inferring this would be one that would do an unbounded number of inference passes. Our inference algorithm uses a fixed number of passes for performance reasons and as a result can't always infer cases like this. We're not against tweaking it to infer more, but this would need to be weighed against perf cost.

Basically TL;DR not a defect (`unknown` inference is sound) but if a good PR existed to fix this case, we'd take it.

**Comment by lynndylanhurley** [comment]
Thanks @RyanCavanaugh ! With that information, I think I can find another approach to my design. I really appreciate your help 🙏

**Comment by lynndylanhurley** [comment]
One more question about this. Using this example:

```ts
const foo2 = createFoo({
  a: {
    i: 1,
  },
  b: {
    ii: (f) => 1 // f is a recursive reference to the base createFoo method param
  }
});
```

I understand why the type of `b.ii` is unknown. But shouldn't the system be able to infer something like this?

```ts
declare const foo2: Foo<{
    a: {
        i: unknown;
    };
    b: {
        ii: unknown;
    };
}>;
```

Currently the entire `b` object is inferred as `unknown` (i.e. no `b.ii` key), and it's making it difficult for me to find a workaround:

```ts
declare const foo2: Foo<{
    a: {
        i: unknown;
    };
    b: unknown;
}>;
```

Is there another way to structure this so that the nested keys are inferred correctly?
